### ISSUE TITLE
Update handler for JSON fields 

### ISSUE THREAD

**Issue by philmetzger** [marceljuenemann/rxdb-supabase#13]
Hey, I was looking into the default updateHandler you created and for my usecase I would always be thrown into the Error:
https://github.com/marceljuenemann/rxdb-supabase/blob/main/src/supabase-replication.ts#L260

I suspect because my id field is a uuid in supabase?

So I created my own updateHandler:
```
push: {
        updateHandler: async (row: RxReplicationWriteToMasterRow<any>) => {
          let query = supabase.from(collections.dumps.name).update(row.newDocumentState, { count: 'exact' });
          Object.entries(row.assumedMasterState!).forEach(([field, value]) => {
            if (field === 'id') {
              query = query.eq(field, value);
            }
          });
          const { error, count } = await query;
          if (error) throw error;
          return count == 1;
        },
      },
```

And I just wanted to see what you think?
Mainly im not too sure why yours is so complicated compared to just using the id always?
Do you foresee any issues here?

**Comment by marceljuenemann** [comment]
The error you linked to is thrown if the *JavaScript* type is unsupported, i.e. if you are trying to store an array or object instead of a string or number. It doesn't know about the database type. UUID should be no problem if it's a string in JavaScript.

The default update handler only updates a row if all the fields match the expected state, so if the row was modified by a different client (e.g. a different user or on a different device changed the state while the current client was offline), it will not be updated. Your handler will always update the row, so if your app goes back online, it might override any changes done in the meantime.

**Comment by marceljuenemann** [comment]
The reason the library doesn't support arrays is just that AFAIK there isn't an easy way to compare equality with the supabase client for array or JSON columns.

If you write your own update handler, you can either create your own logic to determine the new values of the rows (e.g. merging tags?), or just ignore the tags field (i.e. check for field === 'tags')

The README of this project should certainly give more guidance on this. Maybe there could also be an option to ignore certain fields, so when you pass the tags field, it would ignore the field for the equality checks

**Comment by marceljuenemann** [comment]
Well, RxDB also knows what the state before the update was locally. If there was no conflicting update from a different client, then that state should be the same as the remote state. Hence this is called assumedMasterState.

The defaultUpdateHandler generates a query like this:
UPDATE foo
SET a=newState.a, b=newState.b
WHERE a=assumedMasterState.a AND b=assumedMasterState.b

So the problem here is just that I didn't know how to do b=assumedMasterState.b if b is an array or JSON object column

**Comment by colin-chadwick** [comment]
Hi, 

I had the same problem today and wanted to share my findings. 

This should do it:

```ts
private async defaultUpdateHandler(
    row: RxReplicationWriteToMasterRow<RxDocType>,
  ): Promise<boolean> {
    let query = this.options.client
      .from(this.table)
      .update(row.newDocumentState, {
        count: 'exact',
      })

    Object.entries(
      row.assumedMasterState!,
    ).forEach(([field, value]) => {
      const type = typeof value
      if (type === 'string' || type === 'number') {
        query = query.eq(field, value)
      } else if (type === 'boolean' || value === null) {
        query = query.is(field, value)
      } else if (type === 'object') {
        query = query.containedBy(field, value)
      } else {     
        throw new Error(`replicate: Unsupported field of type ${type}`)
      } 
    })
    const { error, count } = await query
    if (error) throw error
    return count === 1
  }
```

See: https://supabase.com/docs/reference/javascript/containedby?example=on-jsonb-columns


Best,
Colin

**Comment by marceljuenemann** [comment]
@colin-chadwick interesting, that certainly looks like a step in the right direction! `containedBy` doesn't check for equality though, but rather whether all the elements in the database field are also in the `value` passed. So there could be cases where the value was modified in the database, yet we still override those changes. Which kind of breaks the contract of the `defaultUpdateHandler` (although this might be fine in your case or other people's cases).

I guess if we used both `containedBy` and `contains` together, that would actually check for equality though :) I'd like to look more into how this handles nested objects, and whether it actually is an equality check, but there might be a general solution here.

**Comment by colin-chadwick** [comment]
You're right! It works for my case, but I think when this is combined with `contains`, this could be a general solution. I'll keep you posted in case I find something helpful. 🙂

**Comment by marceljuenemann** [comment]
So `containedBy` plus `contains` translates to `A <@ B AND A @> B` in Postgres, which isn't the same as equals because it does not compare the ordering of array elements ([see here](https://stackoverflow.com/a/58802265/1620264)). I think that's a problem because it's not uncommon to use a JSON field to store the ordering of some elements.

That said, if you use a `jsonb` field instead of `json`, a simple `=` operator actually does what we want in Postgres. So I think the problem when I wrote this code was just that the supabase client did not correctly translate JSON values into an `eq.` Postgres query. This might just be solved by now, or it should be possible to solve with a pull request to supabase.
